src/test/java/com/google/devtools/build/lib/remote/FakeExecutionService.java

```diff
@@ -0,0 +1,176 @@
+// Copyright 2020 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.google.devtools.build.lib.remote;
+
+import build.bazel.remote.execution.v2.ExecuteRequest;
+import build.bazel.remote.execution.v2.ExecuteResponse;
+import build.bazel.remote.execution.v2.ExecutionGrpc.ExecutionImplBase;
+import build.bazel.remote.execution.v2.WaitExecutionRequest;
+import com.google.common.collect.ImmutableList;
+import com.google.longrunning.Operation;
+import com.google.protobuf.Any;
+import com.google.rpc.Code;
+import com.google.rpc.Status;
+import io.grpc.stub.StreamObserver;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+public class FakeExecutionService extends ExecutionImplBase {
+
+  private final OperationProvider executeOperationProvider = new OperationProvider();
+  private final OperationProvider waitExecutionOperationProvider = new OperationProvider();
+  private int execTimes;
+  private int waitTimes;
+
+  public static class OperationProvider {
+
+    private final Map<String, Deque<ImmutableList<Supplier<Operation>>>> operationProvider =
```

Comment:
nit: add a comment as to what this data structure does

AFAIU it's a map from the request to the list of operations to be returned for each instance of that request, with `Supplier` used for either throwing an exception or returning an `Operation`.

Reply:
Done by copying your words :)